Add documentation

hofaflo opened the pull request:
This PR serves to discuss the introduction of documentation pages.

To-Do (links lead to the rendered version of the current state of this PR):
- [x] [Home](https://mnelab--281.org.readthedocs.build/en/281/index.html)
- [x] [Tutorial](https://mnelab--281.org.readthedocs.build/en/281/tutorial/index.html)
  - [x] Installation
  - [x] First steps
- [x] Examples (step-by-step guides for specific use cases)
  - [x] [ERDS](https://mnelab--281.org.readthedocs.build/en/281/examples/erds_maps.html)
  - [x] [ERP](https://mnelab--281.org.readthedocs.build/en/281/examples/erp.html)
- [x] [About](https://mnelab--281.org.readthedocs.build/en/281/about.html)
- [x] link in main readme

Two questions so far:
- What additional information should be included in "first steps"?
- The [MNE-Python tutorial](https://mne.tools/stable/auto_tutorials/evoked/30_eeg_erp.html) for ERPs uses an MEG data set which is part of MNE-sample-data, so it requires downloading a 1.5GB tar.gz. @cbrnr can you suggest an ERP data set (preferably few files and manually downloadable)?

cbrnr commented:
This looks really great already! I'll think about if anything is missing from the "first steps" page. Maybe we should ask someone completely new to MNELAB?

Re ERP data, maybe you can find something on the [BNCI Horizon 2020](http://bnci-horizon-2020.eu/database/data-sets) website? Dataset 12 (visual P300 speller) looks promising, but the data is stored in `.mat` files. To use it, we need to support importing from `.mat`, but this will be not that easy because the data can be structured arbitrarily.

Maybe the [ERP Core datasets](https://osf.io/thsqg/) are better? I'd use a P3 one, they seem to be available in the EEGLAB format.

Small suggestion for the docs website: can you remove the footer "ads" (i.e. leave only the Copyright part)?

cbrnr commented:
OK, let's disable as much as we can. You are referring to [this commit](https://github.com/pradyunsg/furo/commit/e1bda78483fac80fb2e465b5187c926abfd02afb), right? I wonder why this decision was made though - for aesthetic reasons I would prefer to have as few words as possible in the footer. Can this be customized somehow (i.e. not with setting some variables, but maybe providing a custom HTML template)?

hofaflo commented:
> You are referring to [this commit](https://github.com/pradyunsg/furo/commit/e1bda78483fac80fb2e465b5187c926abfd02afb), right?

Exactly

> Can this be customized somehow (i.e. not with setting some variables, but maybe providing a custom HTML template)?

Yes, with a custom template that's relatively easy, see the current version.

hofaflo commented:
The text part of the ERDS maps example is more or less done, I'll add screenshots after the relevant PRs are merged :)

cbrnr commented on doc/source/examples/erds_maps.md:
```diff
@@ -1,11 +1,59 @@
+Drag the selected data sets to the `Destination` panel and confirm with `OK`.
+A new entry appears in the sidebar: `S001R06 (appended)`.
+The entries `Samples` and `Length` in the info panel confirm that it is the result of the concatenation.
+To close the original data sets, select them in the sidebar and go to `File -> Close` (or press <kbd>Ctrl</kbd>+<kbd>F4</kbd>).
```
On macOS, the shortcut is <kbd>⌘</kbd>+<kbd>w</kbd>.

hofaflo replied:
Oh, then I'd not mention shortcuts at all, people will see them in the menu anyway.

cbrnr commented:
Can/should we add an "About" section (aka "Impressum" in German)? I would at least want to mention Sphinx and the Furo theme there (since we don't have it in the footer anymore). If you have ideas what else to put there please let me know!

hofaflo commented:
Yes, the sphinx and furo attributions are a good idea! I've created the about page, but have no idea what else to put there so far 😅

cbrnr commented on doc/source/examples/erds_maps.md:
```diff
@@ -0,0 +1,59 @@
+Splitting the raw data into epochs requires _events_.
+The info panel shows that the data set already contains _annotations_, but no events.
+To convert annotations to events, select `Tools -> Create events from annotations`.
+The `Events` entry in the info panel now reads `90 (1, 2, 3)`, indicating there are a total of 90 events, with IDs 1, 2, and 3:
```
```suggestion
The `Events` entry in the info panel now reads `90 (1, 2, 3)`, indicating there are a total of 90 events with IDs 1, 2, and 3:
```

cbrnr commented on doc/source/examples/erds_maps.md:
```diff
@@ -0,0 +1,59 @@
+In the sidebar, select `S001R06`, then go to `Edit -> Append data...`.
+Select `S001R10` and `S001R14` in the `Source` panel - to select multiple items, hold <kbd>Ctrl</kbd> (<kbd>⌘</kbd> on macOS) while clicking.
+Drag the selected data sets to the `Destination` panel and confirm with `OK`.
+A new entry appears in the sidebar: `S001R06 (appended)`.
```
Maybe rename this data set to "S001" in the sidebar?

cbrnr commented on doc/source/examples/erds_maps.md:
```diff
@@ -0,0 +1,59 @@
+ERDS maps can only be calculated from "Epochs".
+
+## Pick channels
+Considering the motor imagery tasks, electrodes C3, Cz, and C4, should show the most activity.
```
```suggestion
Considering the motor imagery tasks, electrodes C3, Cz, and C4 should show the strongest activity.
```

cbrnr commented on doc/source/examples/erds_maps.md:
```diff
@@ -0,0 +1,59 @@
+
+## Pick channels
+Considering the motor imagery tasks, electrodes C3, Cz, and C4, should show the most activity.
+Go to `Edit -> Pick channels...`, and select the required channels.
```
```suggestion
Go to `Edit -> Pick channels...` and select the desired channels.
```

cbrnr commented on doc/source/examples/erds_maps.md:
```diff
@@ -0,0 +1,59 @@
+## Pick channels
+Considering the motor imagery tasks, electrodes C3, Cz, and C4, should show the most activity.
+Go to `Edit -> Pick channels...`, and select the required channels.
+Don't worry about the channel names containing trailing dots.
```
```suggestion
Don't worry about the channel names containing trailing dots, that's how they are stored in the original data files.
```

cbrnr commented on doc/source/examples/erds_maps.md:
```diff
@@ -0,0 +1,59 @@
+Don't worry about the channel names containing trailing dots.
+
+## Plot ERDS maps
+Select `Plot -> ERDS maps...` and enter the following configuration values:
```
```suggestion
Select `Plot -> ERDS maps...` and enter the following values:
```

cbrnr commented on doc/source/examples/erds_maps.md:
```diff
@@ -0,0 +1,59 @@
+
+Check `Significance mask`, leave `alpha` at `0.05` and confirm with `OK`.
+Calculating the significance masks can take about 30 seconds.
+Two figures open up, one per event type.
```
```suggestion
Eventually, two figures open up, one per event type.
```

cbrnr commented on doc/source/examples/erp.md:
```diff
@@ -0,0 +1,6 @@
+# Event Related Potentials (ERPs)
+
+Based on [this MNE-Python tutorial](https://mne.tools/stable/auto_tutorials/evoked/30_eeg_erp.html).
```
```suggestion
This example is based on [this MNE-Python tutorial](https://mne.tools/stable/auto_tutorials/evoked/30_eeg_erp.html).
```

hofaflo commented on doc/source/examples/erds_maps.md:
```diff
@@ -0,0 +1,59 @@
+- [S001R14.edf](https://physionet.org/files/eegmmidb/1.0.0/S001/S001R14.edf?download)
+
+## Load data sets
+In MNELAB, go to `File -> Open...`, select the three downloaded files and click `Open`.
```
Good point! How would you write filenames?

cbrnr replied:
I would italicize file names, that's a convention I've seen in many technical books.

hofaflo commented on doc/source/examples/erds_maps.md:
```diff
@@ -0,0 +1,59 @@
+## Plot ERDS maps
+Select `Plot -> ERDS maps...` and enter the following configuration values:
+- Frequency range: `2` to `36`
+- Step size: `1`
```
50s vs 25s on my not-too-new laptop, so I guess 0.5 is fine!

hofaflo commented:
Thanks for your feedback, hope I fixed everything :)

I also added a draft of the ERP example, menu screenshots are postponed until #295 is resolved.

cbrnr commented:
Unfortunately the rendered arrow looks ugly, it's not in the middle of the line. Is it because it is bold and italic? If not, maybe we should use a simple encash instead (surrounded by &nbsp; maybe?).

hofaflo commented:
Not highlighting the arrow doesn't change too much 🤷 

![image](https://user-images.githubusercontent.com/22592497/153942516-ec1902f8-6095-4eb3-b64d-c43648d2c76d.png)

I'll try to find a better one

hofaflo commented:
Which option do you like best?

![image](https://user-images.githubusercontent.com/22592497/153943580-9e5187c0-2af9-4988-aa8c-570cecc12bf5.png)

cbrnr commented:
The second or fifth ones look great! WDYT?

hofaflo commented:
The second one it is!

cbrnr commented:
Let's maybe use a plain old endash –
<img width="877" alt="Screen Shot 2022-02-15 at 08 12 04" src="https://user-images.githubusercontent.com/4377312/154010988-fe4a1e3c-5fb2-4ab3-9e2d-f88f8a3951e5.png">

cbrnr commented:
General comment: IMO the headings are too large. Could we increase the level of each heading so that they are a little smaller? Or should we rather adapt the template?

cbrnr commented on doc/source/examples/erp.md:
```diff
@@ -0,0 +1,91 @@
+![evoked potentials event 1](./images/erp/evoked_event_1.png) ![evoked potentials event 3](./images/erp/evoked_event_3.png)
+
+## Plot evoked topomaps
+Now we will create topomaps of the potentials evoked by the auditory event (ID 1) at -200 ms, 100 ms, and 400 ms.
```
```suggestion
Now we will create topomaps of the potentials evoked by the auditory event "1" at -200 ms, 100 ms, and 400 ms.
```

cbrnr commented on doc/source/examples/erp.md:
```diff
@@ -0,0 +1,91 @@
+
+## Plot evoked topomaps
+Now we will create topomaps of the potentials evoked by the auditory event (ID 1) at -200 ms, 100 ms, and 400 ms.
+Go to **_Plot&nbsp;–&nbsp;Evoked&nbsp;topomaps..._** and select Event 1.
```
```suggestion
Go to **_Plot&nbsp;–&nbsp;Evoked&nbsp;topomaps..._** and select event "1".
```

cbrnr commented on doc/source/examples/erp.md:
```diff
@@ -0,0 +1,91 @@
+## Compare conditions
+So far, we have always plotted individual channels, with different events in separate figures.
+To compare the auditory event to the visual one, go to **_Plot&nbsp;–&nbsp;Evoked comparison_**.
+Judging by the first topomap (at 0.093 s) in the [joint plot above](#joint-plots), we expect a large negative peak in frontal-central positions (FC) for event 1.
```
```suggestion
Judging by the first topomap (at 0.093 s) in the [joint plot above](#joint-plots), we expect a large negative peak in frontal-central positions (FC) for event "1".
```

cbrnr commented on doc/source/examples/erp.md:
```diff
@@ -0,0 +1,91 @@
+So far, we have always plotted individual channels, with different events in separate figures.
+To compare the auditory event to the visual one, go to **_Plot&nbsp;–&nbsp;Evoked comparison_**.
+Judging by the first topomap (at 0.093 s) in the [joint plot above](#joint-plots), we expect a large negative peak in frontal-central positions (FC) for event 1.
+The [channel location plot](#plot-channel-locations) tells us that the relevant channels are EEG 010–014.
```
```suggestion
The [channel location plot](#plot-channel-locations) informs us that the relevant channels are EEG 010–014.
```

cbrnr commented on doc/source/examples/erp.md:
```diff
@@ -0,0 +1,91 @@
+So in the dialog, select channels EEG 010, EEG 011, EEG 012, EEG 013, and EEG 014.
+Leave both event types selected, but change **_Combine&nbsp;channels_** to **_mean_**.
+After confirming with **_OK_**, you will see the figure below.
+Each line represents the average over the selected channels for a single event type, and shaded areas represent 95% confidence intervals.
```
```suggestion
Each line represents the average over the selected channels for a single event type, and shaded ribbons represent 95% confidence intervals.
```

cbrnr commented:
Another oddity of the website: the text "Compute and visualize ERDS maps" in the left sidebar wraps differently depending on whether it is selected (bold) or not. Can we change this or at least work around it (by e.g. finding a new title which doesn't wrap)?

cbrnr commented:
And another suggestion: On the homepage, maybe let's put the first paragraph of text before the screenshots? The links to the Tutorial and Examples can remain below the figure.

Finally, can you make the MNELAB logo in the sidebar a little smaller?

cbrnr commented:
Sorry, new things keep popping up. You mentioned a favicon today, and I thought you meant an app icon. In fact, I would love a custom favicon, and I just had an idea: can we use the default RTD icon, but use red and blue colors for the background of the two pages? The two colors red and blue should be picked from the MNELAB logo. Of course this is only possible if their default logo is under a permissive license.

hofaflo commented:
> General comment: IMO the headings are too large. Could we increase the level of each heading so that they are a little smaller? Or should we rather adapt the template?

Level increases are apparently ignored and I can't seem to figure out how to use a custom css/sass template :/

> Another oddity of the website: the text "Compute and visualize ERDS maps" in the left sidebar wraps differently depending on whether it is selected (bold) or not. Can we change this or at least work around it (by e.g. finding a new title which doesn't wrap)?

For me it always wraps 😅 Making the sidebar wider is apparently [not intended](https://github.com/pradyunsg/furo/blob/d0f880da9453d6a5c2363b9b5c24927ed0c2747f/src/furo/assets/styles/variables/_layout.scss). So yes, title change... "ERDS maps" and "Event Related Potentials"? 

> And another suggestion: On the homepage, maybe let's put the first paragraph of text before the screenshots?

Sure, do you prefer whitespace or a title to keep the text from being too far at the top?

> Finally, can you make the MNELAB logo in the sidebar a little smaller?

The logo in general and/or the top/bottom padding?

> if their default logo is under a permissive license.

looks like [maybe](https://read-the-docs-guidelines.readthedocs-hosted.com/branding.html#logo-guidelines), I'm not sure what this means for their favicon

cbrnr commented:
> Level increases are apparently ignored and I can't seem to figure out how to use a custom css/sass template :/

That's a bummer, but let's postpone this - can you create an issue so that we don't forget?

> For me it always wraps 😅 Making the sidebar wider is apparently [not intended](https://github.com/pradyunsg/furo/blob/d0f880da9453d6a5c2363b9b5c24927ed0c2747f/src/furo/assets/styles/variables/_layout.scss). So yes, title change... "ERDS maps" and "Event Related Potentials"?

Yes!

> Sure, do you prefer whitespace or a title to keep the text from being too far at the top?

A title or nothing at all.

> The logo in general and/or the top/bottom padding?

The logo is too large.

> looks like [maybe](https://read-the-docs-guidelines.readthedocs-hosted.com/branding.html#logo-guidelines), I'm not sure what this means for their favicon

Good enough, but let's make sure to mention that we use a modified version in the About page?

cbrnr commented:
Regarding fonts, other sites based on Furo use headings that are not boldface - this looks much better (e.g. https://pip.pypa.io/en/stable/user_guide/). Did you change anything from the defaults or do we have to adapt the theme so that headings are thinner?

hofaflo commented:
> > looks like [maybe](https://read-the-docs-guidelines.readthedocs-hosted.com/branding.html#logo-guidelines), I'm not sure what this means for their favicon
> 
> Good enough, but let's make sure to mention that we use a modified version in the About page?

You aren't worried about the "do not alter our logos in any way" part? 🤔 

> Regarding fonts, other sites based on Furo use headings that are not boldface - this looks much better (e.g. https://pip.pypa.io/en/stable/user_guide/). Did you change anything from the defaults or do we have to adapt the theme so that headings are thinner?

Note to self: always look at the RTD-rendered version, not the local one. I initially though you were talking about font size 😅 The boldface was added in a recent [commit](https://github.com/pradyunsg/furo/commit/3dc84c0f648d7407996810dafe7bd502b4ea50d7) and is part of the [2022.02.14 release](https://pradyunsg.me/furo/changelog/#jazzy-jasmine). So I vote for pinning furo to 2022.01.02 until we figure out how to modify the style template.

cbrnr commented:
> You aren't worried about the "do not alter our logos in any way" part? 🤔

1. I wouldn't say their favicon is their logo.
2. It's CC-BY licensed, which means we are free to modify it. They are contradicting this license when they restrict modifications.
3. If these two arguments are not convincing, we could easily come up with a favicon that looks similar and has the two colors.

> Note to self: always look at the RTD-rendered version, not the local one. I initially though you were talking about font size 😅 The boldface was added in a recent [commit](https://github.com/pradyunsg/furo/commit/3dc84c0f648d7407996810dafe7bd502b4ea50d7) and is part of the [2022.02.14 release](https://pradyunsg.me/furo/changelog/#jazzy-jasmine). So I vote for pinning furo to 2022.01.02 until we figure out how to modify the style template.

Yes, I was talking about font size originally, but then I noticed that it's really the boldface that makes the headings look so much larger. Yes, let's pin the version, and maybe you can open an issue in the Furo repo and ask how this can be customized/reverted?

hofaflo commented:
This is ready from my point of view :)

cbrnr commented on mnelab/mainwindow.py:
```diff
@@ -259,10 +259,13 @@ def __init__(self, model: Model):
         self.actions["about"] = help_menu.addAction("&About", self.show_about)
         self.actions["about_qt"] = help_menu.addAction("About &Qt", self.show_about_qt)
-
+        self.actions["documentation"] = help_menu.addAction(
```
I think a separator before this line would look good.

cbrnr commented:
Docs are live! Thanks a ton @hofaflo!
